### DISCUSSION TITLE
Extension point to add extra hidden fields?

### DISCUSSION THREAD

**Discussion by wimdeblauwe** [thymeleaf/thymeleaf#934] [Q&A]
Is there an extension point to add extra hidden fields. For issue https://github.com/wimdeblauwe/htmx-spring-boot-thymeleaf/issues/36 I would like to add an CSRF hidden field like Thymeleaf itself does when using a form. I already have an [HtmxAttributeProcessor](https://github.com/wimdeblauwe/htmx-spring-boot-thymeleaf/blob/main/src/main/java/io/github/wimdeblauwe/hsbt/thymeleaf/HtmxAttributeProcessor.java), but I am unsure how I could add an extra hidden field from that code.

Any pointers would be highly appreciated.

**Comment by landsman** [comment]
Bump. Is there anyone with the info? 🙏

**Comment by ultraq** [comment]
If you have your own attribute processor, then you'll have access to an instance of an [`IElementTagStrucureHandler`](https://www.thymeleaf.org/apidocs/thymeleaf/3.1.2.RELEASE/org/thymeleaf/processor/element/IElementTagStructureHandler.html) which will let you insert any "model" (ie: HTML elements) into the HTML template being processed with the `insertBefore` and `insertImmediatelyAfter` methods.

As for creating a model that represents some HTML, you can go about it by getting an [`IModelFactory`](https://www.thymeleaf.org/apidocs/thymeleaf/3.1.2.RELEASE/org/thymeleaf/model/IModelFactory.html) instance (usually from the template context that is also passed to your attribute processor via `templateContext.getModelFactory()`) and using its methods to build one, eg:

```groovy
// To add something like <input type="hidden" name="csrftoken" value="...">
// after the element being processed by your tag processor
var modelFactory = templateContext.getModelFactory()
var inputModel = modelFactory.createStandaloneElementTag("input",
  Map.of("type", "hidden", "name", "csrftoken", "value", ...),
  AttributeValueQuotes.DOUBLE, false, false
)
structureHandler.insertImmediatelyAfter(inputModel, false)
```

There are a lots of other methods in those classes for creating and inserting models, which you can find in the javadoc links I've provided above.
